be/src/storage/binlog_builder.cpp

```diff
@@ -0,0 +1,346 @@
+        if (!st.ok()) {
+            current_active_writer->close(false);
+            LOG(WARNING) << "Fail to reset active writer when discarding data for version " << version
+                         << ", file path: " << current_active_writer->file_path() << ", " << st;
+        } else {
+            // reuse the writer
+            reused_writer = result.active_writer;
+        }
+    } else if (prev_active_writer != nullptr) {
+        if (fail_delete_new_files) {
+            // 2.2 just resize the file but not reuse it, and refer to BinlogBuilder#abort
+            // step 2.2.1 for the reason
+            st = FileSystemUtil::resize_file(prev_active_writer->file_path(), params->active_file_meta->file_size());
+            LOG_IF(WARNING, !st.ok()) << "Fail to resize the active file when discarding data for version: " << version
+                                      << ", file path: " << prev_active_writer->file_path() << ", " << st;
+        } else {
+            // 2.3 previous writer already closed, and reopen it for reuse
+            StatusOr<BinlogFileWriterPtr> status_or = BinlogFileWriter::reopen(
+                    prev_active_writer->file_id(), prev_active_writer->file_path(), params->max_page_size,
+                    params->compression_type, params->active_file_meta.get());
+            if (!status_or.ok()) {
+                LOG(WARNING) << "Fail to reopen writer when discarding data for version " << version
+                             << ", file path: " << prev_active_writer->file_path() << ", " << st;
+            } else {
+                // reuse the writer
+                reused_writer = status_or.value();
+            }
+        }
+    }
+
+    return reused_writer;
+}
+
+Status BinlogBuilder::build_duplicate_key(int64_t version, const RowsetSharedPtr& rowset,
+                                          BinlogBuilderParamsPtr& builder_params, BinlogBuildResult* result) {
+    std::shared_ptr<BinlogBuilder> builder =
+            std::make_shared<BinlogBuilder>(version, rowset->creation_time() * 1000000, builder_params);
+    Status status;
+    if (rowset->num_rows() == 0) {
+        status = builder->add_empty();
+    } else {
+        RowsetSegInfo seg_info(version, 0);
+        std::vector<SegmentSharedPtr>& segments = rowset->segments();
+        for (int32_t seg_index = 0; seg_index < rowset->num_segments(); seg_index++) {
+            int num_rows = segments[seg_index]->num_rows();
+            if (num_rows == 0) {
+                continue;
+            }
+            seg_info.seg_index = seg_index;
+            status = builder->add_insert_range(seg_info, 0, num_rows);
+            if (!status.ok()) {
+                LOG(WARNING) << "Fail to add_insert_range for duplicate key, rowset: " << rowset->rowset_id()
+                             << ", segment index: " << seg_index << ", number of rows " << num_rows
+                             << ", version: " << version << ", " << status;
```

Comment:
when possible, add tablet_id in most logs

Reply:
fixed